### DISCUSSION TITLE
(🎁) Support for basedtyping/basedmypy

### DISCUSSION THREAD

**Discussion by KotlinIsland** [pydantic/pydantic#8337] [Question] [answered]
[basedmypy](https://github.com/KotlinIsland/basedmypy)/basedtyping supports additional typing constructs:

```py
# Intersections
i1: Sized & Container[int]
i2: Intersection[Sized, Container[int]]
```

And alternative syntaxes for denoting types:
```py
# bare literal types
a: 1 | 2 | True

# tuple literal types
b: (int, str) = 1, ""
```

And there are others planned:
```py
# variadic tuple types
t: tuple[int | str] = (1, "", 1, "")

# string literal types
s: "before" | "after" | "plain" | "wrap"

# explicit forward references
f: ForwardReference["A"]
class A: ...

# Removal of promoted/duck types
a: float = 1  # error: int does not extend float
```

# So
We could categorize theses features depending on if they affect other typing things:
- non breaking:
    - `basedtyping.Intersection`
    - tuple literal types
    - bare literals
    - explicit forward references
- breaking:
    - based variadic tuple types
    - string literal types
    - removal of promoted types
- Not all of these can exist without `__future__.annotations` enabled (`A & B`), and would need a custom type annotation parser to work. An implementation of this is WIP, but not yet available.

It would be great if Pydantic could support these features as much as possible.

My question is: is there any possibility of this being done? I would assume the Pydantic project wouldn't want to maintain support for non-standard functionality. So would this be possible with a Pydantic plugin?

**Comment by lmmx** [comment]
Based on what?

**Comment by Viicos** [comment]
> My question is: is there any possibility of this being done? I would assume the Pydantic project wouldn't want to maintain support for non-standard functionality. So would this be possible with a Pydantic plugin?

I'm assuming this can be implemented as a third party package, using [custom types](https://docs.pydantic.dev/latest/concepts/types/#custom-types).

**Reply by KotlinIsland** [reply]
so i can't write a plugin that will add support for `basedtyping.Intersection`, i need to modify `Intersection` to include type metadata?

**Reply by Viicos** [reply]
Ideally you would provide an [`Annotated[basedtyping.Intersection, ...]` type alias](https://docs.pydantic.dev/latest/concepts/types/#as-an-annotation), but this isn't always ideal. We are thinking about adding a way to easily customize support for custom types, e.g. via a registry:

```python
class Model(BaseModel):
    t: basedtyping.Intersection

    model_config = {"custom_types": {basedtyping.Interesection: some_hook}}
```

Or by customizing the `GenerateSchema` class. However, we are still in the process of tweaking a lot of things on this class and the API isn't public for the moment. This might come in the future.

**Reply by KotlinIsland** [reply, marked as answer]
https://github.com/KotlinIsland/pydantic-basedtyping doesn't support `Intersection` (yet)

**Reply by Viicos** [reply]
Interesting, however note that we don't provide any API stability guarantees on internal modules. In your case, you are monkeypatching `eval_type_lenient` which is deprecated.

**Reply by KotlinIsland** [reply]
yeah, i know it's not optimal, but the plugin system didn't provide any means to intercept or manipulate type annotations

**Reply by Viicos** [reply]
I'm currently working on a `GenerateSchema` refactor, and I'm trying to doing so in a way that would allow some kind of hook system so that a given arbitrary type `<type>` can work, without having to define the `__get_pydantic_core_schema__` methods. I don't know if this will support intercepting the type evaluation though.

**Reply by KotlinIsland** [reply]
that's awesome, could you please link to it so I can subscribe and keep up to date?

**Reply by Viicos** [reply]
For now probably https://github.com/pydantic/pydantic/issues/10297
